core/src/main/scala/org/apache/spark/ui/jobs/UIData.scala

```diff
@@ -112,9 +112,9 @@ private[spark] object UIData {
   /**
    * These are kept mutable and reused throughout a task's lifetime to avoid excessive reallocation.
    */
-  class TaskUIData private(
-      private var _taskInfo: TaskInfo,
-      private var _metrics: Option[TaskMetricsUIData]) {
+  class TaskUIData private(private var _taskInfo: TaskInfo) {
+
+    private[this] var _metrics: Option[TaskMetricsUIData] = Some(TaskMetricsUIData.EMPTY)
```

Comment:
The only way for this to become `None` is if `updateTaskMetrics` is called with `None`.

`updateTaskMetrics` is called in two places:

- In JobProgressListener.onTaskEnd, where the metrics are from `Option(taskEnd.taskMetrics)`, where `taskEnd.taskMetrics` can be `null` in case the task has failed (according to docs).
- In JobProgressListener.onExecutorMetricsUpdate, where the metrics are guaranteed to be defined / non-None.